diff --git a/dnsmasq.conf.example b/dnsmasq.conf.example
index 6b00bd3..adb2b18 100644
--- a/dnsmasq.conf.example
+++ b/dnsmasq.conf.example
@@ -25,6 +25,13 @@
 #conf-file=%%PREFIX%%/share/dnsmasq/trust-anchors.conf
 #dnssec
 
+# Specify file containing spurious IP. dnsmasq will ignore DNS
+# replies with these IP.
+#spurious-ip-file=/etc/spurious_ips.conf
+# Or specify IP address one by one.
+#spurious-ip = 64.33.99.47
+#spurious-ip = 4.36.66.178
+
 # Replies which are not DNSSEC signed may be legitimate, because the domain
 # is unsigned, or may be forgeries. Setting this option tells dnsmasq to
 # check that an unsigned reply is OK, by finding a secure proof that a DS 
@@ -535,7 +542,7 @@
 # if there is one.
 #dhcp-script=/bin/echo
 
-# Set the cachesize here.
+# Set the cachesize here, maximum value 65535.
 #cache-size=150
 
 # If you want to disable negative caching, uncomment this.
@@ -548,6 +555,10 @@
 # seconds) here.
 #local-ttl=
 
+# Set a minimum TTL value for entries in the cache.
+# For example min-cache-ttl=300 (ie: 5 minutes).
+#min-cache-ttl=
+
 # If you want dnsmasq to detect attempts by Verisign to send queries
 # to unregistered .com and .net hosts to its sitefinder service and
 # have dnsmasq instead return the correct NXDOMAIN response, uncomment
diff --git a/src/cache.c b/src/cache.c
index 2c3a498..b78ddcc 100644
--- a/src/cache.c
+++ b/src/cache.c
@@ -464,6 +464,8 @@ struct crec *cache_insert(char *name, struct all_addr *addr,
       /* Don;t mess with TTL for DNSSEC records. */
       if (daemon->max_cache_ttl != 0 && daemon->max_cache_ttl < ttl)
 	ttl = daemon->max_cache_ttl;
+      if (daemon->min_cache_ttl != 0 && daemon->min_cache_ttl > ttl)
+	ttl = daemon->min_cache_ttl;
     }
 
   /* if previous insertion failed give up now. */
diff --git a/src/config.h b/src/config.h
index 145820a..59137b3 100644
--- a/src/config.h
+++ b/src/config.h
@@ -200,6 +200,14 @@ RESOLVFILE
 #   endif
 #endif
 
+#ifndef SPURIOUS_IPS
+#   if defined(__uClinux__)
+#      define SPURIOUS_IPS "/etc/config/spurious_ips.conf"
+#   else
+#      define SPURIOUS_IPS "/etc/spurious_ips.conf"
+#   endif
+#endif
+
 #ifndef RUNFILE
 #   if defined(__ANDROID__)
 #      define RUNFILE "/data/dnsmasq.pid"
diff --git a/src/dnsmasq.h b/src/dnsmasq.h
index e74b15a..88dbecb 100644
--- a/src/dnsmasq.h
+++ b/src/dnsmasq.h
@@ -543,6 +543,13 @@ struct resolvc {
   char *name;
 };
 
+struct spurious_ip {
+  u8 (*ip)[4];
+  int max_entry; /* maximum entries to hold in ip array. */
+  int nr_entry;
+  char *name;
+};
+
 /* adn-hosts parms from command-line (also dhcp-hostsfile and dhcp-optsfile */
 #define AH_DIR      1
 #define AH_INACTIVE 2
@@ -576,6 +583,9 @@ struct hostsfile {
 #define FREC_DO_QUESTION       64
 #define FREC_ADDED_PHEADER    128
 #define FREC_CHECK_NOSIGN     256
+#define FREC_RECURSIVE        512
+
+#define FREC_RECURSIVE_BIT      9
 
 #ifdef HAVE_DNSSEC
 #define HASH_SIZE 20 /* SHA-1 digest size */
@@ -900,6 +910,7 @@ extern struct daemon {
 
   unsigned int options, options2;
   struct resolvc default_resolv, *resolv_files;
+  struct spurious_ip spurious;
   time_t last_resolv;
   char *servers_file;
   struct mx_srv_record *mxnames;
@@ -933,7 +944,7 @@ extern struct daemon {
   int max_logs;  /* queue limit */
   int cachesize, ftabsize;
   int port, query_port, min_port;
-  unsigned long local_ttl, neg_ttl, max_ttl, max_cache_ttl, auth_ttl;
+  unsigned long local_ttl, neg_ttl, max_ttl, min_cache_ttl, max_cache_ttl, auth_ttl;
   struct hostsfile *addn_hosts;
   struct dhcp_context *dhcp, *dhcp6;
   struct ra_interface *ra_interfaces;
@@ -1135,6 +1146,7 @@ int legal_hostname(char *c);
 char *canonicalise(char *s, int *nomem);
 unsigned char *do_rfc1035_name(unsigned char *p, char *sval);
 void *safe_malloc(size_t size);
+void *safe_realloc(void *p, size_t size);
 void safe_pipe(int *fd, int read_noblock);
 void *whine_malloc(size_t size);
 int sa_len(union mysockaddr *addr);
@@ -1162,6 +1174,7 @@ int read_write(int fd, unsigned char *packet, int size, int rw);
 
 int wildcard_match(const char* wildcard, const char* match);
 int wildcard_matchn(const char* wildcard, const char* match, int num);
+int cmp_int32(const void *a, const void *b);
 
 /* log.c */
 void die(char *message, char *arg1, int exit_code);
diff --git a/src/forward.c b/src/forward.c
index 2cf29eb..ae41235 100644
--- a/src/forward.c
+++ b/src/forward.c
@@ -337,6 +337,9 @@ static int forward_query(int udpfd, union mysockaddr *udpaddr,
 	  memcpy(forward->hash, hash, HASH_SIZE);
 	  forward->forwardall = 0;
 	  forward->flags = 0;
+	  if (header->hb3 & HB3_RD)
+	    forward->flags |= FREC_RECURSIVE;
+	  /* my_syslog(LOG_WARNING, _("Forward original HB3_RD: %d"), header->hb3 & HB3_RD); */
 	  if (norebind)
 	    forward->flags |= FREC_NOREBIND;
 	  if (header->hb4 & HB4_CD)
@@ -529,6 +532,7 @@ static size_t process_reply(struct dns_header *header, time_t now, struct server
   char **sets = 0;
   int munged = 0, is_sign;
   size_t plen; 
+  int extract_addr_ret = 0;
 
   (void)ad_reqd;
   (void) do_bit;
@@ -623,8 +627,13 @@ static size_t process_reply(struct dns_header *header, time_t now, struct server
 	  SET_RCODE(header, NOERROR);
 	  cache_secure = 0;
 	}
-      
-      if (extract_addresses(header, n, daemon->namebuff, now, sets, is_sign, check_rebind, no_cache, cache_secure, &doctored))
+
+        /* store the return value for later use
+           the special return value -1 is used to denote spurious IP */
+        extract_addr_ret = extract_addresses(header, n, daemon->namebuff, now,
+                                             sets, is_sign, check_rebind,
+                                             no_cache, cache_secure, &doctored);
+        if (extract_addr_ret)
 	{
 	  my_syslog(LOG_WARNING, _("possible DNS-rebind attack detected: %s"), daemon->namebuff);
 	  munged = 1;
@@ -666,6 +675,8 @@ static size_t process_reply(struct dns_header *header, time_t now, struct server
       header->nscount = htons(0);
       header->arcount = htons(0);
     }
+  if (extract_addr_ret == -1)
+      return (size_t)-1;
   
   /* the bogus-nxdomain stuff, doctor and NXDOMAIN->NODATA munging can all elide
      sections of the packet. Find the new length here and put back pseudoheader
@@ -673,6 +684,8 @@ static size_t process_reply(struct dns_header *header, time_t now, struct server
   return resize_packet(header, n, pheader, plen);
 }
 
+#define MAX_RETRY 8
+
 /* sets new last_server */
 void reply_query(int fd, int family, time_t now)
 {
@@ -682,14 +695,18 @@ void reply_query(int fd, int family, time_t now)
   union mysockaddr serveraddr;
   struct frec *forward;
   socklen_t addrlen = sizeof(serveraddr);
-  ssize_t n = recvfrom(fd, daemon->packet, daemon->packet_buff_sz, 0, &serveraddr.sa, &addrlen);
+  ssize_t n;
   size_t nn;
   struct server *server;
   void *hash;
+  int retry_cnt = 0;
 #ifndef HAVE_DNSSEC
   unsigned int crc;
 #endif
 
+retry:
+  n = recvfrom(fd, daemon->packet, daemon->packet_buff_sz, 0, &serveraddr.sa, &addrlen);
+
   /* packet buffer overwritten */
   daemon->srv_save = NULL;
   
@@ -780,6 +797,17 @@ void reply_query(int fd, int family, time_t now)
     {
       int check_rebind = 0, no_cache_dnssec = 0, cache_secure = 0;
 
+      /* the recursive desired bit has been changed by mitm */
+      if (((forward->flags & FREC_RECURSIVE) >> FREC_RECURSIVE_BIT)
+          ^ (header->hb3 & HB3_RD)) {
+        my_syslog(LOG_WARNING, _("flags: %d, hb3_rd: %d"),
+                  (forward->flags & FREC_RECURSIVE), (header->hb3 & HB3_RD));
+        if (retry_cnt++ < MAX_RETRY)
+          goto retry;
+        else
+          return;
+      }
+
       if (option_bool(OPT_NO_REBIND))
 	check_rebind = !(forward->flags & FREC_NOREBIND);
       
@@ -993,9 +1021,12 @@ void reply_query(int fd, int family, time_t now)
       else
 	header->hb4 &= ~HB4_CD;
       
-      if ((nn = process_reply(header, now, server, (size_t)n, check_rebind, no_cache_dnssec, cache_secure,
-			      forward->flags & FREC_AD_QUESTION, forward->flags & FREC_DO_QUESTION, 
-			      forward->flags & FREC_ADDED_PHEADER, forward->flags & FREC_HAS_SUBNET, &forward->source)))
+      nn = process_reply(header, now, server, (size_t)n, check_rebind, no_cache_dnssec, cache_secure,
+                         forward->flags & FREC_AD_QUESTION, forward->flags & FREC_DO_QUESTION, 
+                         forward->flags & FREC_ADDED_PHEADER, forward->flags & FREC_HAS_SUBNET, &forward->source);
+      if ((nn == (size_t)-1) && (retry_cnt++ < MAX_RETRY))
+          goto retry;
+      if (nn)
 	{
 	  header->id = htons(forward->orig_id);
 	  header->hb4 |= HB4_RA; /* recursion if available */
diff --git a/src/option.c b/src/option.c
index b08e98e..b3a9c03 100644
--- a/src/option.c
+++ b/src/option.c
@@ -147,6 +147,9 @@ struct myoption {
 #define LOPT_LOCAL_SERVICE 335
 #define LOPT_DNSSEC_TIME   336
 #define LOPT_LOOP_DETECT   337
+#define LOPT_SPURIOUS_FILE 338
+#define LOPT_SPURIOUS_IP   339
+#define LOPT_MINCTTL       340
 
 #ifdef HAVE_GETOPT_LONG
 static const struct option opts[] =  
@@ -253,6 +256,7 @@ static const struct myoption opts[] =
     { "dhcp-broadcast", 2, 0, LOPT_BROADCAST },
     { "neg-ttl", 1, 0, LOPT_NEGTTL },
     { "max-ttl", 1, 0, LOPT_MAXTTL },
+    { "min-cache-ttl", 1, 0, LOPT_MINCTTL },
     { "max-cache-ttl", 1, 0, LOPT_MAXCTTL },
     { "dhcp-alternate-port", 2, 0, LOPT_ALTPORT },
     { "dhcp-scriptuser", 1, 0, LOPT_SCRIPTUSR },
@@ -299,6 +303,8 @@ static const struct myoption opts[] =
     { "quiet-dhcp6", 0, 0, LOPT_QUIET_DHCP6 },
     { "quiet-ra", 0, 0, LOPT_QUIET_RA },
     { "dns-loop-detect", 0, 0, LOPT_LOOP_DETECT },
+    { "spurious-ip-file", 2, 0, LOPT_SPURIOUS_FILE },
+    { "spurious-ip", 1, 0, LOPT_SPURIOUS_IP },
     { NULL, 0, 0, 0 }
   };
 
@@ -457,6 +463,8 @@ static struct {
   { LOPT_QUIET_RA, OPT_QUIET_RA, NULL, gettext_noop("Do not log RA."), NULL },
   { LOPT_LOCAL_SERVICE, OPT_LOCAL_SERVICE, NULL, gettext_noop("Accept queries only from directly-connected networks"), NULL },
   { LOPT_LOOP_DETECT, OPT_LOOP_DETECT, NULL, gettext_noop("Detect and remove DNS forwarding loops"), NULL },
+  { LOPT_SPURIOUS_FILE, ARG_DUP, "<path>", gettext_noop("Specify path to spurious-ip file."), NULL },
+  { LOPT_SPURIOUS_IP, ARG_DUP, "<ipaddr>", gettext_noop("Specify spurious ip."), NULL },
   { 0, 0, NULL, NULL, NULL }
 }; 
 
@@ -698,6 +706,29 @@ static void do_usage(void)
     }
 }
 
+static void add_spurious_ip(char *s)
+{
+  /* skip empty string. */
+  char *token = strtok(s, ".");
+  if (!token || strpbrk(s, " \t\n\r") != NULL)
+    return;
+
+  struct spurious_ip *sp = &daemon->spurious;
+  if (sp->ip == NULL || sp->nr_entry == sp->max_entry)
+    {
+      sp->max_entry *= 2;
+      sp->ip = safe_realloc(sp->ip, sp->max_entry * sizeof(sp->ip[0]));
+    }
+
+  int j = 0;
+  do
+    {
+      sp->ip[sp->nr_entry][j++] = atoi(token);
+      /* my_syslog(LOG_INFO, _("%d"), atoi(token)); */
+    } while ((token = strtok(NULL, ".")) && j < 4);
+  sp->nr_entry++;
+}
+
 #define ret_err(x) do { strcpy(errstr, (x)); return 0; } while (0)
 
 char *parse_server(char *arg, union mysockaddr *addr, union mysockaddr *source_addr, char *interface, int *flags)
@@ -2372,8 +2403,8 @@ static int one_opt(int option, char *arg, char *errstr, char *gen_err, int comma
 	    
 	    if (size < 0)
 	      size = 0;
-	    else if (size > 10000)
-	      size = 10000;
+	    else if (size > 65535)
+	      size = 65535;
 	    
 	    daemon->cachesize = size;
 	  }
@@ -2424,6 +2455,7 @@ static int one_opt(int option, char *arg, char *errstr, char *gen_err, int comma
     case 'T':         /* --local-ttl */
     case LOPT_NEGTTL: /* --neg-ttl */
     case LOPT_MAXTTL: /* --max-ttl */
+    case LOPT_MINCTTL: /* --min-cache-ttl */
     case LOPT_MAXCTTL: /* --max-cache-ttl */
     case LOPT_AUTHTTL: /* --auth-ttl */
       {
@@ -2434,6 +2466,8 @@ static int one_opt(int option, char *arg, char *errstr, char *gen_err, int comma
 	  daemon->neg_ttl = (unsigned long)ttl;
 	else if (option == LOPT_MAXTTL)
 	  daemon->max_ttl = (unsigned long)ttl;
+	else if (option == LOPT_MINCTTL)
+	  daemon->min_cache_ttl = (unsigned long)ttl;
 	else if (option == LOPT_MAXCTTL)
 	  daemon->max_cache_ttl = (unsigned long)ttl;
 	else if (option == LOPT_AUTHTTL)
@@ -3853,6 +3887,20 @@ static int one_opt(int option, char *arg, char *errstr, char *gen_err, int comma
 	break;
       }
 #endif
+
+    case LOPT_SPURIOUS_FILE:
+      daemon->spurious.name = opt_string_alloc(arg);
+      break;
+
+    case LOPT_SPURIOUS_IP:
+      i = daemon->spurious.nr_entry;
+      add_spurious_ip(arg);
+      my_syslog(LOG_INFO, _("add spurious ip %d.%d.%d.%d\n"),
+        daemon->spurious.ip[i][0],
+        daemon->spurious.ip[i][1],
+        daemon->spurious.ip[i][2],
+        daemon->spurious.ip[i][3]);
+      break;
 		
     default:
       ret_err(_("unsupported option (check that dnsmasq was compiled with DHCP/TFTP/DNSSEC/DBus support)"));
@@ -4182,7 +4230,24 @@ void read_servers_file(void)
   
   read_file(daemon->servers_file, f, LOPT_REV_SERV);
 }
- 
+
+void read_spurious_ip_file(char *buff)
+{
+  char *line = NULL;
+  FILE *f = NULL;
+
+  if (!(f = fopen(daemon->spurious.name, "r")))
+    {
+      my_syslog(LOG_ERR, _("cannot read %s: %s"), daemon->spurious.name, strerror(errno));
+      return;
+    }
+
+  while ((line = fgets(buff, MAXDNAME, f)))
+    add_spurious_ip(line);
+
+  my_syslog(LOG_INFO, _("load %d spurious ip"), daemon->spurious.nr_entry);
+  fclose(f);
+}
 
 #ifdef HAVE_DHCP
 void reread_dhcp(void)
@@ -4296,6 +4361,7 @@ void read_opts(int argc, char **argv, char *compile_opts)
   daemon->dhcp_server_port = DHCP_SERVER_PORT;
   daemon->default_resolv.is_default = 1;
   daemon->default_resolv.name = RESOLVFILE;
+  daemon->spurious.max_entry = 32;
   daemon->resolv_files = &daemon->default_resolv;
   daemon->username = CHUSER;
   daemon->runfile =  RUNFILE;
@@ -4525,6 +4591,16 @@ void read_opts(int argc, char **argv, char *compile_opts)
   if (daemon->if_names || daemon->if_except || daemon->if_addrs || daemon->authserver)
     reset_option_bool(OPT_LOCAL_SERVICE); 
 
+  if (daemon->spurious.name != NULL)
+    read_spurious_ip_file(buff);
+
+  struct spurious_ip *sp = &daemon->spurious;
+  /* release extra memory allocated */
+  sp->ip = safe_realloc(sp->ip, sp->nr_entry * sizeof(sp->ip[0]));
+  sp->max_entry = sp->nr_entry;
+  /* sort ip for bsearch */
+  qsort(sp->ip, sp->nr_entry, sizeof(sp->ip[0]), cmp_int32);
+
   if (testmode)
     {
       fprintf(stderr, "dnsmasq: %s.\n", _("syntax check OK"));
diff --git a/src/rfc1035.c b/src/rfc1035.c
index 8a7d260..dcb6c94 100644
--- a/src/rfc1035.c
+++ b/src/rfc1035.c
@@ -892,6 +892,12 @@ static int find_soa(struct dns_header *header, size_t qlen, char *name, int *doc
   return minttl;
 }
 
+static int is_spurious(struct in_addr addr) {
+  struct spurious_ip *sp = &daemon->spurious;
+  return bsearch(&addr.s_addr, sp->ip, sp->nr_entry,
+  	sizeof(sp->ip[0]), cmp_int32) != NULL;
+}
+
 /* Note that the following code can create CNAME chains that don't point to a real record,
    either because of lack of memory, or lack of SOA records.  These are treated by the cache code as 
    expired and cleaned out that way. 
@@ -1109,6 +1115,15 @@ int extract_addresses(struct dns_header *header, size_t qlen, char *name, time_t
 			}
 #endif
 		      
+		      /* before insert it into cache, check if it is a fake addr */
+                      if (is_spurious(addr.addr.addr4)) {
+                        my_syslog(LOG_WARNING, _("spurious IP: %d.%d.%d.%d"),
+                                  ((u8 *)&addr.addr.addr4)[0],
+                                  ((u8 *)&addr.addr.addr4)[1],
+                                  ((u8 *)&addr.addr.addr4)[2],
+                                  ((u8 *)&addr.addr.addr4)[3]);
+                        return -1;
+                      }
 		      newc = cache_insert(name, &addr, now, attl, flags | F_FORWARD | secflag);
 		      if (newc && cpp)
 			{
diff --git a/src/util.c b/src/util.c
index a729f33..8d816aa 100644
--- a/src/util.c
+++ b/src/util.c
@@ -245,6 +245,17 @@ void *safe_malloc(size_t size)
   return ret;
 }    
 
+/* for use during startup */
+void *safe_realloc(void *p, size_t size)
+{
+  void *ret = realloc(p, size);
+  
+  if (!ret)
+    die(_("realloc could not get memory"), NULL, EC_NOMEM);
+     
+  return ret;
+}    
+
 void safe_pipe(int *fd, int read_noblock)
 {
   if (pipe(fd) == -1 || 
@@ -656,3 +667,15 @@ int wildcard_matchn(const char* wildcard, const char* match, int num)
 
   return (!num) || (*wildcard == *match);
 }
+
+int cmp_int32(const void *a, const void *b)
+{
+  const int32_t *ia = (int32_t *)a;
+  const int32_t *ib = (int32_t *)b;
+
+  if (*ia > *ib)
+      return 1;
+  if (*ia == *ib)
+      return 0;
+  return -1;
+}
